src/compiler/checker.ts

```diff
@@ -14597,6 +14599,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             }
         }
         const props = propSet ? arrayFrom(propSet.values()) : [singleProp];
+        const id = `${checkFlags & CheckFlags.ReadPartial ? 'p' : ''}${skipObjectFunctionPropertyAugment ? 's' : ''}${isUnion ? 'u' : 'i'}${getSymbolListId(props)}`;
+        const cached = cachedUnionOrIntersectionPropertySymbols.get(id);
+        if (cached) {
+            return cached;
+        }

```

Comment:
It's worth explaining why I'm even caching those...

The problem that it solves is that [`getTypeWithThisArgument` call](https://github.dev/microsoft/TypeScript/blob/efc9c065a2caa52c5bebd08d730eed508075a78a/src/compiler/checker.ts#L13075-L13076) for the `Bar1` creates a **new** intersection type, one that doesn't have any reference (through `.target` or similar) to the "original" intersection (the base type).

So `createUnionOrIntersectionProperty` creates 2 separate synthetic properties for `a` for both the base type and the derived type (even though they both refer to the same symbols, declarations, etc). That, in turn, fails [`derived === base` check](https://github.dev/microsoft/TypeScript/blob/efc9c065a2caa52c5bebd08d730eed508075a78a/src/compiler/checker.ts#L44697-L44700) that is responsible for checking if the property was implemented or not.

An alternative solution to this caching could be extending this `derived === base` with some extra conditions.